src/meta/meta_bulk_load_ingestion_context.cpp

```diff
@@ -0,0 +1,204 @@
+DSN_TAG_VARIABLE(bulk_load_node_min_disk_count, FT_MUTABLE);
+
+ingestion_context::ingestion_context() { reset_all(); }
+
+ingestion_context::~ingestion_context() { reset_all(); }
+
+// ingestion_context::partition_node_info
+void ingestion_context::partition_node_info::create(const partition_configuration &config,
+                                                    const config_context &cc)
+{
+    pid = config.pid;
+    std::unordered_set<rpc_address> current_nodes;
+    current_nodes.insert(config.primary);
+    for (const auto &secondary : config.secondaries) {
+        current_nodes.insert(secondary);
+    }
+    for (const auto &node : current_nodes) {
+        std::string disk_tag;
+        if (cc.get_disk_tag(node, disk_tag)) {
+            node_disk[node] = disk_tag;
+        }
+    }
+}
+
+// ingestion_context::node_context
+void ingestion_context::node_context::init_disk(const std::string &disk_tag)
+{
+    if (disk_count.find(disk_tag) != disk_count.end()) {
+        return;
+    }
+    disk_count[disk_tag] = 0;
+}
+
+int32_t ingestion_context::node_context::get_max_disk_ingestion_count(
+    const int32_t max_node_ingestion_count) const
+{
+    FAIL_POINT_INJECT_F("ingestion_node_context_disk_count", [](string_view count_str) -> int32_t {
+        auto count = 0;
+        buf2int32(count_str, count);
+        return count;
+    });
+
+    auto node_disk_count = disk_count.size() > FLAGS_bulk_load_node_min_disk_count
+                               ? disk_count.size()
+                               : FLAGS_bulk_load_node_min_disk_count;
+    if (max_node_ingestion_count % node_disk_count == 0) {
+        return max_node_ingestion_count / node_disk_count;
+    }
+    return max_node_ingestion_count / node_disk_count + 1;
+}
+
+bool ingestion_context::node_context::check_if_add(const std::string &disk_tag)
+{
+    auto max_node_ingestion_count = FLAGS_bulk_load_node_max_ingesting_count;
+    if (node_count >= max_node_ingestion_count) {
+        dwarn_f("node[{}] has {} partition executing ingestion, max_count = {}",
+                address.to_string(),
+                node_count,
+                max_node_ingestion_count);
+        return false;
+    }
+
+    if (disk_count.find(disk_tag) == disk_count.end()) {
+        disk_count[disk_tag] = 0;
+    }
+    auto max_disk_ingestion_count = get_max_disk_ingestion_count(max_node_ingestion_count);
+    if (disk_count[disk_tag] >= max_disk_ingestion_count) {
+        dwarn_f("node[{}] disk[{}] has {} partition executing ingestion, max_count = {}",
+                address.to_string(),
+                disk_tag,
+                disk_count[disk_tag],
+                max_disk_ingestion_count);
+        return false;
+    }
+    return true;
+}
+
+void ingestion_context::node_context::add(const std::string &disk_tag)
+{
+    disk_count[disk_tag]++;
+    node_count++;
+}
+
+void ingestion_context::node_context::decrease(const std::string &disk_tag)
+{
+    node_count--;
```

Comment:
People may understand `node_count` means `count of node`, is better named `node_ingesting_count` ?

The follow is same,  `disk_ingesting_count`?

Reply:
Those belongs to `ingestion_context` class, and they won't be accessed out of this class, I think adding `ingesting` for their name to a little bit duplicated.